82 refactor color ssn utility for performance and code reuse

nilsoberg opened the pull request.

nilsoberg committed:
Create Nextflow script stubs for sharing Nextflow code with new pipelines

nilsoberg committed:
Create modular color SSN workflow

nilsoberg committed:
Cluster and color SSNs

- Update Annotations modules for reusable field name constant
- Perl scripts for extracting edgelist and coloring XGMML
- Python script for computing clusters

nilsoberg committed:
Update color SSN nextflow configuration to reflect changes

nilsoberg committed:
Update Python and Perl requirements for building Docker container

nilsoberg committed:
Retrieve ID lists for color SSN/GNT/cluster analysis tools

nilsoberg committed:
Retrieve FASTA sequences from ID lists.

Required a refactor of the est/import code by moving the files from the pipelines/est/import/lib to the root/lib directory.

nilsoberg committed:
Create annotated mapping table

The annotated mapping tables are now created in the color SSN pipeline.  The two
tables that are current created are the mapping table which contains metadata
for every UniProt ID (expanded from UniRef/RepNode IDs), and a second table
containing a list of IDs that are SwissProt sequences with the description.
A color mapping file is now created by color_xgmml.pl. Also, since the
pipeline requires access to a FASTA database to retrieve the sequences,
the test was updated to reflect that.

nilsoberg committed:
Add missing annotation script

nilsoberg committed:
Handle singletons when writing XGMML

nilsoberg committed:
Create numbering and mapping files for clusters and singletons

Several steps are needed to obtain the information for the pipeline (and future
derivatives).  During the ID list retrieval process, cluster sizes are computed.
During the cluster computation process, singletons are saved as well as information
on the relationship between clusters numbered by number of sequences vs number
of nodes.

nilsoberg committed:
Centralize command line options and shared functions

The command line argument parsing was centralized into EFI::Options.
Instances of shared functions were moved to the EFI::SSN::Util::ID module.

nilsoberg committed:
Compute convergence ratio stub script

nilsoberg committed:
Compute Color SSN pipeline stats

nilsoberg committed:
Fix bug in network cluster computation

nilsoberg committed:
Color SSN docker config supports file system structure better

nilsoberg committed:
Fix errors in library path and command line arguments

nilsoberg committed:
Bring Python code in line with project standards

nilsoberg committed:
Update RepNode field constants.

Legacy CD-HIT fields in EFI::Annotations are removed.
Constant renamed and updated for RepNode cluster ID list.

nilsoberg committed:
Order cluster number header

nilsoberg committed:
Fix issue with handling metanodes

nilsoberg committed:
Fix duplicate UniRef ID issue

nilsoberg committed:
Add new fields to EFI::Annotations for color SSN

nilsoberg committed:
Refactor XGMML writing code into external module files

nilsoberg committed:
Add POD for modules

nilsoberg committed:
Fix bugs after further testing

nilsoberg committed:
Add documentation for Perl scripts

1ndy commented on docs/source/pipelines/shared/python/compute_clusters.rst:
```diff
@@ -0,0 +1,28 @@
+================
+
+Clusters are computed based on node connectivity using igraph. The input is a simple
+edgelist file for performance reasons (i.e. the graph isn't contructed on the
```
"contructed"

1ndy commented on pipelines/colorssn/colorssn.nf:
```diff
@@ -1,65 +1,8 @@
+include { color_and_retrieve } from "../shared/nextflow/color_workflow.nf"
 
 workflow {
-    if (params.ssn_input =~ /\.zip/) {
-        ssn_file = unzip_input(params.ssn_input)
-    } else {
-        ssn_file = params.ssn_input
-    }
```
Would it be possible to test for the magic number (`PK\x03\x04` I think) as well as the extension? I'm just thinking about a user misnaming their file, possibly maliciously

1ndy commented on pipelines/shared/nextflow/color_workflow.nf:
```diff
@@ -0,0 +1,202 @@
+    """
+}
+
+process get_fasta {
+    publishDir params.final_output_dir, mode: 'copy'
+    input:
+        path id_list_dir
+    output:
+        path 'fasta/'
+    """
+    mkdir fasta
+    mkdir fasta/uniprot_ids fasta/uniref90_ids fasta/uniref50_ids
+    for id_file in $id_list_dir/uniprot_ids/*.txt; do
+        [ -e "\${id_file}" ] || continue
+        fasta_file=\${id_file/%.txt/.fasta}
+        fasta_file=\${fasta_file/#$id_list_dir/fasta}
+        perl $projectDir/../shared/perl/get_sequences.pl --fasta-db ${params.fasta_db} --sequence-ids-file \${id_file} --output-sequence-file \${fasta_file}
+    done
+    for id_file in $id_list_dir/uniref90_ids/*.txt; do
+        [ -e "\${id_file}" ] || continue
+        fasta_file=\${id_file/%.txt/.fasta}
+        fasta_file=\${fasta_file/#$id_list_dir/fasta}
+        perl $projectDir/../shared/perl/get_sequences.pl --fasta-db ${params.fasta_db} --sequence-ids-file \${id_file} --output-sequence-file \${fasta_file}
+    done
+    for id_file in $id_list_dir/uniref50_ids/*.txt; do
+        [ -e "\${id_file}" ] || continue
+        fasta_file=\${id_file/%.txt/.fasta}
+        fasta_file=\${fasta_file/#$id_list_dir/fasta}
+        perl $projectDir/../shared/perl/get_sequences.pl --fasta-db ${params.fasta_db} --sequence-ids-file \${id_file} --output-sequence-file \${fasta_file}
+    done
+    """
```
I would try to make `get_fasta` accept a tuple of the path to the idlist and the verison tag. In the `workflow` section, you could then filter for files that exist and aggregate by version tag and pass all of the id files for that version to the process or pass every pair of (file, version) to `get_fasta` to parallelize all of the lookups. Either way I would try to avoid the explicit for loops and manual directory management. I can elaborate on this if needed.

1ndy replied:
https://www.nextflow.io/docs/latest/channel.html#operators

1ndy commented on pipelines/shared/nextflow/color_workflow.nf:
```diff
@@ -0,0 +1,202 @@
+    """
+}
+
+// This is necessary to avoid Docker mounting the same directory twice; this occurs
+// because we need to mount the root of the project so that we can get access to the
+// lib directory. When using a SSN that is inside the project directory then the project
+// directory is also mounted.
```
Is this mainly an issue in testing? If so could the tests be restructured to avoid the added complexity

nilsoberg replied:
I didn't think about it, but yes, this is only an issue when the test data is inside the source code directory tree.  I didn't like my change and I feel it's a hack, so I'll remove this copy stuff and figure out something else for the tests.  The latter change will come in a separate issue/PR.

1ndy commented on pipelines/shared/nextflow/color_workflow.nf:
```diff
@@ -0,0 +1,202 @@
+        path 'id_lists/', emit: 'id_lists'
+        path 'cluster_sizes.txt', emit: 'cluster_sizes'
+    """
+    mkdir id_lists
+    mkdir id_lists/uniprot_ids id_lists/uniref90_ids id_lists/uniref50_ids
+    perl $projectDir/../shared/perl/get_id_lists.pl --cluster-map $cluster_id_map --singletons $singletons \
+        --uniprot id_lists/uniprot_ids --uniref90 id_lists/uniref90_ids --uniref50 id_lists/uniref50_ids \
+        --seqid-source-map $seqid_source_map --cluster-sizes cluster_sizes.txt \
+        --config ${params.efi_config} --db-name ${params.efi_db}
```
Scripts should still be runnable without Nextflow, could you make Perl create these directories instead of doing it in the process? You could also output tuples instead of directory names so that each file is tagged with its uniref version.

1ndy commented on pipelines/shared/nextflow/color_workflow.nf:
```diff
@@ -0,0 +1,202 @@
+    output:
+        path 'conv_ratio.txt', emit: 'conv_ratio'
+    """
+    perl $projectDir/../shared/perl/compute_conv_ratio.pl --cluster-map $cluster_id_map --index-seqid-map $index_seqid_map \
+        --edgelist $edgelist --seqid-source-map $seqid_source_map --conv-ratio conv_ratio.txt
+    """
+}
+
+process get_cluster_stats {
+    publishDir params.final_output_dir, mode: 'copy'
+    input:
+        path cluster_id_map
+        path seqid_source_map
+        path singletons
+    output:
+        path 'stats.txt', emit: 'stats'
+    """
+    perl $projectDir/../shared/perl/compute_stats.pl --cluster-map $cluster_id_map --seqid-source-map $seqid_source_map \
+        --singletons $singletons --stats stats.txt
```
`conv_ratio` and `compute_stats` both exist as python functions. I think they both accept parquet files so they obviously wouldn't work for this but in the future it would be good to define those computations in terms of nodes and edges exactly once and then only write code to adapt the input type to work with those functions. The calculations could be Perl or Python, the important thing would be the ability to read diverse file types.

1ndy commented on pipelines/shared/python/compute_clusters.py:
```diff
@@ -0,0 +1,82 @@
+
+
+def get_node_id_mapping(id_map_file: str) -> tuple[dict, dict]:
+
```
Add a docstring explaining the file format and what the mappings represent

1ndy commented on pipelines/shared/python/compute_clusters.py:
```diff
@@ -0,0 +1,82 @@
+            if len(parts) >= 2:
+                idx = int(parts[0])
+                idx_seqid_map[idx] = parts[1]
+                idx_node_size[idx] = int(parts[2]) if len(parts) >= 3 and parts[2] else 1
```
stylistically I would avoid these kind of one-liner if statements, I don't see a lot of python code that uses them. Also, add a comparison for `parts[2]` to make it clear what you expect it to be

1ndy commented on pipelines/shared/python/network.py:
```diff
@@ -0,0 +1,181 @@
+            # The dict idx here represents the old (self.components) cluster number; sorting
+            # in the file will be done by sequence number
+            for idx in self.cluster_num_map_by_seq:
+                cluster_num_seq = self.cluster_num_map_by_seq[idx]
+                cluster_num_node = self.cluster_num_map_by_node[idx]
+                cluster_size_seq = self.cluster_size_by_seq[idx]
+                cluster_size_node = self.cluster_size_by_node[idx]
+                fh.write(f"{cluster_num_seq}\t{cluster_size_seq}\t{cluster_num_node}\t{cluster_size_node}\n")
```
I don't have a strong suggestion on this but since everything is going to end up in a TSV, is there a way to construct the data as a table rather than in dictionaries? I might have missed something but for the most part it looks like you aren't doing random insertions/lookups so constructing parallel arrays and inserting into a pandas dataframe or arrow table might retain more of the semantic meaning of the data as its being processed. Just a thought.

nilsoberg replied:
I'm not exactly happy with the code since it's more of a brute force approach (although I don't consider it a hack).  Because the same clusters are ordered differently depending on metanode+cluster size or just cluster size I am storing the values in separate dictionaries.  I agree it could be different, but I need to move on to other things.  If I have to refactor this code in the future I'll take your suggestions into account.

nilsoberg committed:
Check input zip file magic number

nilsoberg committed:
Tests require external test data to be stored outside of source directory tree

nilsoberg committed:
Fix typo

nilsoberg committed:
Add docstring and clarify code

nilsoberg committed:
Refactor Nextflow workflow to improve FASTA retrieval.

Docker config is updated to mitigate duplicate Docker mount points.
ID list directories are created by the Perl script.

nilsoberg committed:
Update color pipeline to fix Nextflow paradigm

ID lists and FASTA are retrieved using tuples and in parallel
Replaced use of single quotes with double quotes for consistency

1ndy commented on pipelines/shared/python/compute_clusters.py:
```diff
@@ -47,6 +47,22 @@ def create_parser() -> argparse.ArgumentParser:
     return parser
 
 
+"""
+Load a file that maps node index to sequence ID and sequence metanode size
+(e.g. the file has node_idx, seq_id, seq_size for column names). If the
+input network is UniProt (i.e. not UniRef or RepNode) then the size column
+will always be zero.
+
+Parameters
+----------
+    id_map_file
+        path to the file that contains the mapping
+
+Returns
+-------
+    dict mapping node index to sequence ID
+    dict mapping node index to metanode size
+"""
```
this should go underneath the function signature

nilsoberg committed:
Fix docstring placement

nilsoberg committed:
Fix test path

nilsoberg committed:
Merge branch 'nextflow-test' into 82-refactor-color-ssn-utility-for-performance-and-code-reuse
